### ISSUE TITLE
Wiimote (with Wii Motion Plus) controller support

### ISSUE THREAD

**Issue by Andrope** [carl-anders/slimevr-wrangler#7]
Would it be possible to add support for Wiimotes that have Wii Motion Plus? They have a gyroscope, although I don't know how you'd implement their support in Rust as the current input interpreters are only in C, Java, Python... (maybe with a C -> Rust thing?). I'm no programmer so I'm not qualified to know if it's doable, but I'd love to see this feature as I have 2 Wiimotes almost begging me to be used for Full Body Tracking.

Wiimote libraries : http://www.wiibrew.org/wiki/Wiimote/Library
Wiimote motion analysis (only for the accelerometer tho) : http://www.wiibrew.org/wiki/Wiimote/Motion_analysis

Thanks in advance!

**Comment by MisutaaAsriel** [comment]
I'm equally interested in this. 2 Joy-Cons alone are not very useful for FBT, but if I I could throw a Wiimote or two in the mix then perhaps it'd be more feasible!

**Comment by alexmusa** [comment]
There is what seems to be significant work for this made by @cesmec here https://github.com/cesmec/slimevr-wrangler/tree/wiimote-hack but one of the dependency is in a private repo of theirs.

**Comment by cesmec** [comment]
Yes, I've started working on this as I like the idea and have a few different Wiimotes lying around. The results seem promising so far, but there are some things that I want to change or at least try out in the other repo before making it public. Hopefully I have some time to work on it over the next few weeks.

**Comment by kaiboisdumb** [comment]
> Yes, I've started working on this as I like the idea and have a few different Wiimotes lying around. The results seem promising so far, but there are some things that I want to change or at least try out in the other repo before making it public. Hopefully I have some time to work on it over the next few weeks.

Hi @cesmec. Any update on your progress with this? I read over [this other repository that had the idea to use Wiimotes](https://github.com/kruemmelbande/SlimeWiir) and they claimed that Wiimotes are awful as trackers, so I'm curious if your findings have changed.

**Comment by cesmec** [comment]
Hi @kaiboisdumb, writing _**the next few weeks**_ in my last comment was a bit too optimistic, but I could actually make some progress on this. My wiimote library is now public and published and I'm working on integrating that version into slimevr-wrangler.

How useful Wiimotes are as trackers: I wouldn't call it awful, but also not too great and I also wouldn't buy a Wiimote for this purpose. In my tests, they seemed to work mostly OK for slower movements, but faster movements resulted in too much motion. Maybe my code for the fast mode is wrong (it's calculated differently than the slow mode) or the sensors might just not be accurate enough for that.

I'll post another update if I find out more about that and will do some more testing, but this will take a bit of time (hopefully less than since my last comment).

**Comment by kruemmelbande** [comment]
> > Yes, I've started working on this as I like the idea and have a few different Wiimotes lying around. The results seem promising so far, but there are some things that I want to change or at least try out in the other repo before making it public. Hopefully I have some time to work on it over the next few weeks.
> 
> Hi @cesmec. Any update on your progress with this? I read over [this other repository that had the idea to use Wiimotes](https://github.com/kruemmelbande/SlimeWiir) and they claimed that Wiimotes are awful as trackers, so I'm curious if your findings have changed.

To be fair, at the time i didnt do fusion properly. Its working okay now, but its still not great. Definitely worse than joycons which i already consider basically unusable, but thats up to your own judgement. The biggest issue ive found is that fast movements are not tracked at all.

**Comment by cesmec** [comment]
Hey, it's been a while but I finally have another update on this.

I could fix a bug in my calculations for the fast mode. However, fast movements are inaccurate and the Wiimotes drift over time. I've also realized that the accuracy depends on the bluetooth adapter of the PC. One that I was using was not able to keep up with the data rate of the Wiimote (which is higher compared to Joy Cons as those send multiple measurements combined into a single update). How well the connection works can be seen by the status of "Healthy" or "Laggy IMU" in the same way as Joy Cons already display.
I've now pushed the changes to https://github.com/cesmec/slimevr-wrangler/tree/wiimote for those who want to try it out.

**Comment by ArturEsp** [comment]
> Hey, it's been a while but I finally have another update on this.
> 
> I could fix a bug in my calculations for the fast mode. However, fast movements are inaccurate and the Wiimotes drift over time. I've also realized that the accuracy depends on the bluetooth adapter of the PC. One that I was using was not able to keep up with the data rate of the Wiimote (which is higher compared to Joy Cons as those send multiple measurements combined into a single update). How well the connection works can be seen by the status of "Healthy" or "Laggy IMU" in the same way as Joy Cons already display. I've now pushed the changes to https://github.com/cesmec/slimevr-wrangler/tree/wiimote for those who want to try it out.

How do you get it to work? I've downloaded the wrangler app from the link that you sent, and it seems to be the same app as in the orginal page. Pressing 1+2 on the controllers doesnt to anything, neither connecting them through the bluetooth settings. Do I need to download something else? @cesmec

**Comment by cesmec** [comment]
@ArturEsp if you've cloned it using git, make sure that you've checked out the wiimote branch.
It should then show the text `Searching for Joycon and Wii controllers...` instead of `Searching for Joycon controllers...`
With the app open, the controllers should then connect when pressing 1+2 without having to add them through the bluetooth settings.

Also, the Wii remotes currently only work on Windows and Linux (no support for MacOS).

**Comment by ArturEsp** [comment]
> [@ArturEsp](https://github.com/ArturEsp) if you've cloned it using git, make sure that you've checked out the wiimote branch. It should then show the text `Searching for Joycon and Wii controllers...` instead of `Searching for Joycon controllers...` With the app open, the controllers should then connect when pressing 1+2 without having to add them through the bluetooth settings.
> 
> Also, the Wii remotes currently only work on Windows and Linux (no support for MacOS).

@cesmec  I dont know what im doing wrong, I am going to your branch ( https://github.com/cesmec/slimevr-wrangler/tree/wiimote ), scolling down into the README, and under the photo, where the Setup steps begin, i click on the second one ( Downnload SlimeVR Wrangler), and it downloads me the same file as in the original post, with no modifications for the Wiiremote support. Do I need to download the files that appear at the start? Sorry if im being a little troublesome, its just that im a noob at github. Thanks.

**Comment by MisutaaAsriel** [comment]
> [@cesmec](https://github.com/cesmec) I dont know what im doing wrong, I am going to your branch ( https://github.com/cesmec/slimevr-wrangler/tree/wiimote ), scolling down into the README, and under the photo, where the Setup steps begin, i click on the second one ( Downnload SlimeVR Wrangler), and it downloads me the same file as in the original post, with no modifications for the Wiiremote support. Do I need to download the files that appear at the start? Sorry if im being a little troublesome, its just that im a noob at github. Thanks.

@ArturEsp The readme links are not updated for that fork (they redirect back to the main project). Looking at the repository, it looks like you need to **build it yourself**, as there's no GitHub Actions artifacts nor Releases (*Artifacts: built/bundled files as part of automated builds; Releases: Manually added builds, milestones, for download*).

**Comment by ArturEsp** [comment]
> > [@cesmec](https://github.com/cesmec) I dont know what im doing wrong, I am going to your branch ( https://github.com/cesmec/slimevr-wrangler/tree/wiimote ), scolling down into the README, and under the photo, where the Setup steps begin, i click on the second one ( Downnload SlimeVR Wrangler), and it downloads me the same file as in the original post, with no modifications for the Wiiremote support. Do I need to download the files that appear at the start? Sorry if im being a little troublesome, its just that im a noob at github. Thanks.
> 
> [@ArturEsp](https://github.com/ArturEsp) The readme links are not updated for that fork (they redirect back to the main project). Looking at the repository, it looks like you need to **build it yourself**, as there's no GitHub Actions artifacts nor Releases (_Artifacts: built/bundled files as part of automated builds; Releases: Manually added builds, milestones, for download_).

Okay thanks! But how exactly do I do that? (Building it myself) @MisutaaAsriel

**Comment by MisutaaAsriel** [comment]
> Okay thanks! But how exactly do I do that? (Building it myself) [@MisutaaAsriel](https://github.com/MisutaaAsriel)

🤷 Beats me. Project has no build instructions. Looking at the GitHub actions YAML, it seems to be built with a tool called "cargo" and is a Rust project. That's all I can infer.
